Move Twig functions and filters to an actual Twig extension.

heino opened the pull request:
#### Issue
Timber currently extends Twig via Twig_Environment::addFunction() and ::addFilter(), which is called from WordPress hooks/actions. This way a function or filter is forcefully injected into Twig once and for all, and it cannot be replaced, or even removed (since calling these methods directly on Twig_Environment takes precedence over any registered extension).

#### Solution
By moving the individual functions and filters into an actual Twig extension, overloading becomes possible. Also, it would be possible to extend the extension class, and filter out or replace individual functions/filters. Twig uses extensions this way internally, which enables users to overload Twig’s internal elements, although this is not recommended.

#### Impact
Currently users of Timber are unable to overload functions and filters used by Timber, so there should be no issues with backward compatibility.

Users will be able to override Timber’s functions and filters, just like Twig. According to the manual overloading the built-in elements is not recommended as it might be confusing.

#### Usage
The proposed extension class allows filtering the function and filters via Wordpress hooks. It is also possible to make further manipulations when extension of the class.

#### Considerations
Order matters when overloading via extensions, and last loaded takes precedence.
The currently proposed modification loads the extension in Loader::get_twig(), right after the cache extension is loaded, and would therefore be easily overloaded. Loading the extension via an action hook (like ‘timber/twig/extensions’) should be considered…

#### Testing
Currently no testing has been done beyond verifying that Timber’s functions and filters is available.

heino committed:
Move Twig functions and filters til actual Twig extension.

jarednova committed:
Merge branch 'master' into extension

jarednova commented:
Awesome, thanks @heino! I'm really glad you're digging deeper into the Twig side of things. Looks like the current build has errors, but keep at it!

heino committed:
Change function and filter arrays into associative arrays (to make it easier to filter out or replace).

heino committed:
Add 3 missing methods required by Twig v1.x

heino committed:
Update test to reflect that Twig::time_ago() was moved to TwigExtension along with the other internal helper functions

heino committed:
Last relevant stuff moved to the Loader class

heino committed:
Fix namespace prefix

heino committed:
Fix callable, which shouldn't have been in an array.

heino committed:
Move filters

heino committed:
Add loader method to extention, and move point of load from Loader::get_twig() Leader to Timber:init()

heino committed:
Move addition of Timber's escapers from constructor to loadIntoEnvironment()

heino committed:
Rename activate() into activateWordpressHook()

This is done to avoid confusion, since the extension could be used
without Timber and Wordpress.

heino committed:
Ups! Removal of unintended test...

jarednova committed:
Merge branch 'master' into extension

jarednova commented:
Hi @heino — thanks so much for your work on this and #1500, I'm really excited to dig in. Let me know when you're ready for me to start closely reviewing this. It looks like you're still pushing commits so I don't want to start that process before you're ready

heino committed:
Move TwigExtension.php to TwigExtension/CoreExtension.php

heino committed:
Change references to TwigExtension to CoreExtension

jarednova committed:
Merge branch 'master' into extension
